include/packetgraph/packet.h

```diff
@@ -60,6 +60,27 @@ unsigned int pg_packet_max_len(pg_packet_t *packet);
  */
 int pg_packet_set_len(pg_packet_t *packet, unsigned int len);
 
+void pg_packet_set_l2_len(pg_packet_t *packet, unsigned int len);
+int pg_packet_get_l2_len(pg_packet_t *packet);
+void pg_packet_set_l3_len(pg_packet_t *packet, unsigned int len);
+int pg_packet_get_l3_len(pg_packet_t *packet);
+void pg_packet_set_l4_len(pg_packet_t *packet, unsigned int len);
+int pg_packet_get_l4_len(pg_packet_t *packet);
+
+/**
+ * iterate on each packets in @pkts base on @mask
+ *
+ * @pkt:	current packet
+ * @it:		current pos
+ */
+#define PG_PACKETS_FOREACH(pkts, mask, pkt, it)				\
+	pg_packet_t *pkt;						\
+	for (uint64_t tmpmask = (mask), it;				\
+	     ((it = __builtin_ctzll(tmpmask)) || 1)			\
+		     && (pkt = (pkts)[it]) && tmpmask;			\
```

Comment:
but a brick should never receive a burst with a null mask, if I patch it here, I'll need to patch every other mask manipulation functions.

this macro work only with gcc + clang, but it's a macro if some drunkard wand to use another compiler they can simply not use it, and the code should compile anyway.


Also, it's not that hard to reimplement: https://code.woboq.org/qt5/qtbase/src/corelib/tools/qalgorithms.h.html